### PR TITLE
fix(checks): Check `vm.max_map_count` using `cat`

### PR DESCRIPTION
**Pull request by jiriks74**
This solves multiple issues:
- We don't need `sudo` to check the value anymore
- Some systems may not have `sysctl` available
  - IDK about desktops, but my Debian server doesn't have this command (for whatever reason)

### CODE REVIEW

**Comment by aclist** [comment]
- Is this approach consistent across systems?
- If the system does not have sysctl available, is there no functional benefit to writing to the `$conf_file`? In other words, assume there is no pre-existing conf file, the map count is too low, and the system doesn't have sysctl. Seems like we'd be prompting the user for sudo but this would do nothing. (Not an issue with this PR itself, just pointing out a general flaw)

N.B.: A lot of distributions have defaulted to updating the map count to a higher threshold, so this check is mostly redundant on modern systems, but not always.

File: dzgui.sh
Comment (aclist) [inline dzgui.sh:425]: This logger line should be irrelevant if the $ct variable is unused

Comment (jiriks74) [reply dzgui.sh:425]: Missed that one. I'll rip it out asap.

**Comment by jiriks74** [comment]
> * Is this approach consistent across systems?

Should be as this should be a file that the Linux kernel itself creates.

> A lot of distributions have defaulted to updating the map count to a higher threshold, so this check is mostly redundant on modern systems, but not always.

I agree. And this is also why I chose to do this as it's really annoying when my PC already has this set up and I either have to always cancel the prompt or let dzgui write the config file (the check isn't working for me and since you're suppressing basically all errors I don't know what's wrong). I don't want to have the config file as I'm setting up everything through nix so I have to fix it in some way.

File: dzgui.sh
Comment (jiriks74) [reply dzgui.sh:425]: Ok, afaik I removed it already?

Comment (aclist) [reply dzgui.sh:425]: It might be a visual bug in the diff on GitHub's side. It looks like line 425 calls the logger to print the contents of `$ct`, but this variable is no longer being used? Sorry if I misread something.

**Comment by aclist** [comment]
> I agree. And this is also why I chose to do this as it's really annoying when my PC already has this set up and I either have to always cancel the prompt or let dzgui write the config file

I agree; I think this is one of the oldest pieces of code right now and being able to skip sudo invocation entirely for the map count check is a great change.

>  (the check isn't working for me and since you're suppressing basically all errors I don't know what's wrong).

IIRC, stdout and stderr were interfering with the way zenity dialogs were being created (some bugs in zenity), so it was being partially suppressed somewhere, but this is hardly necessary anymore since the UI was rewritten from scratch. Can you point me to a specific place where you feel stderr is being suppressed where it shouldn't be?

Edit: offhand, seems likely to be more to do with the fact that the sysctl command is invoked in a subshell.

> I don't want to have the config file as I'm setting up everything through nix so I have to fix it in some way.

DZGUI should catch existing config files and skip a lot of the pre-setup checks if you've previously installed, but this area should probably be looked at again for sprucing up. (I'm prototyping a new setup wizard on the UI side as well.)

File: dzgui.sh
Comment (jiriks74) [reply dzgui.sh]: Oh I see it now. The line numbers are confusing:
image

**Comment by jiriks74** [comment]
> Can you point me to a specific place where you feel stderr is being suppressed where it shouldn't be?

On the top of my head - the `ping` checks. On NixOS `-4` isn't a thing and `dzgui` failed until I went through the code, found the `ping` commands, ran them myself in the terminal and then I knew that `-4` caused the error. I know that `stderr` was redirected at multiple places but this is the one that comes to mind without going through the code again.

> DZGUI should catch existing config files

It does. But in Nix the config files are done in a weird way (compared to FHS distributions). Currently `dzgui` always creates the file for setting `vm.max_map_count` even though it's already set and `sysctl` returns the expected value. I don't know why it fails (can't see any errors) but it does so every time I provide a password the config file is created. And since I try to have everything on my system configured for `nix` (I don't want to use both `nix` and traditional configuration at the same time) so I have to delete it every time.

**Comment by jiriks74** [comment]
I just tested my changes. The setup runs and I don't get a prompt for `sudo` to increase the value.

**Comment by jiriks74** [comment]
Hello! I'm happy to see your account again!

**Comment by jiriks74** [comment]
> While I'm not sure what the use case is for DZGUI on a (headless?) server context, perhaps we should test whether sysctl even exists on the system before proceeding to the last step.

I do not intend to run it on that server. I mostly wanted to test some things on a FHS distribution and this showed me that, again, I can't take things for granted even if they appear standardized.

> Maybe we should stop checking for the dayz.conf file altogether?

Probably. One issue I think might happen is that the `/proc/sys/vm/max_map_count` doesn't get updated until reboot. I have seen this ***somewhere*** so this is something needing a fact check.

> An alternate way of updating the map count parameter on a system without sysctl would be to echo the contents directly into /proc/sys/vm/max_map_count, but I'm not certain if this value would persist through reboots.

AFAIK this wouldn't persist over boot. This should be a *virtual* file created by the kernel - it won't exist on the filesystem when the OS on that filesystem isn't running.

**Comment by aclist** [comment]
> Hello! I'm happy to see your account again!

Hi 👋

>   I do not intend to run it on that server. I mostly wanted to test some things on a FHS distribution and this showed me that, again, I can't take things for granted even if they appear standardized.

Right.

> Probably. One issue I think might happen is that the /proc/sys/vm/max_map_count doesn't get updated until reboot. I have seen this somewhere so this is something needing a fact check.

I think it's highly improbable that the setting is sticky unless the user takes some further action to apply it to the kernel boot params. I just wanted to point out that if there is no sysctl on the system, then writing the dayz.conf file isn't doing anything, here.

**Comment by jiriks74** [comment]
I'm working on the changes and I'll post what I have done so far any minute now.

One question I have for you - do we need to set up the system? Why not make temporary fix (eg. echoing to `/proc/sys/vm/max_map_count`) and notify the user that they need to set their system up? Just send them to some wiki page about this topic or something. Things like the steam deck and anything that's not over a year old ([see article about ubuntu](https://www.gamingonlinux.com/2024/03/ubuntu-2404-increases-vm-max-map-count-for-smoother-linux-gaming/)) should be fine. Ubuntu Wiki is probably the only one we need to add as I doubt *"normies"* will be using debian.

If we drop this then we don't need to solve these edge cases or modify the user's OS configuration. We won't need to things like NixOS or non-SystemD distributions into account as a result of this as `/proc/sys/vm/max_map_count` is from the kernel, not the init system (afaik).

**Comment by jiriks74** [comment]
So `echo` to `/proc/sys/vm/max_map_count` is possible as per [this article](https://www.theshell.guru/how-to-fix-vm-max_map_count-is-too-low-rocky-centos/):

> To fix this temporarily and without having to reboot your server, issue the following command at the terminal:
> ```bash
> echo 9999999 > /proc/sys/vm/max_map_count
> ```

**Comment by aclist** [comment]
On Steam Deck the map count is high enough OOTB.

Hmm, so basically you seem to be saying to dispense with making it sticky, and just update the value ad-hoc at runtime every time they start the application. It's the user's prerogrative to update their map count on their system if they want it to be higher for other purposes.

This seems like an elegant solution. We need to test this on a real setup to make sure that:

- Given a system with a map count < threshold
- And user starts DZGUI
- DZGUI echoes the new map count
- DayZ actually runs without crashing

I guess the repro steps would be:
- Artifically lower your map count
- Make sure that DayZ crashes without intervention from DZGUI in the kernel params
- Next, let DZGUI echo the map count
- See if DayZ runs this time

**Comment by aclist** [comment]
And we can just emit a warning into the logs and put an admonition in the docs.

**Comment by aclist** [comment]
But wait...this goes back to square one:

- If the setting is not sticky, the user is going to have to be forced to invoke sudo every reboot when they use DZGUI
- And we nag them until they update the map count themselves?

This seems like it would be potentially more annoying...

**Comment by jiriks74** [comment]
This would have to wait until the option is really mainstream. Some research needs to be done about this probably.
I was just thinking about some approaches on it now. If you want to add these things to this PR I'd like to do it right the first time rather than creating a second PR few months later to re-fix this.

The original issue I was solving was that on some systems I got this prompt even though the option didn't need to be set as DZGUI didn't get the current value set and I thought that not relying on another tool and parsing it's output when you can just `cat` the value is better.

**Comment by aclist** [comment]
> The original issue I was solving was that on some systems I got this prompt even though the option didn't need to be set as DZGUI didn't get the current value set and I thought that not relying on another tool and parsing it's output when you can just cat the value is better.

No, I agree with that, better to save a step and not raise a sudo dialog if we can just read the contents of the file into memory. Like you said, the count is already high enough on the vast majority of (modernish) distributions.

> If you want to add these things to this PR I'd like to do it right the first time rather than creating a second PR few months later to re-fix this.

Up to you, but it's a small fix to drop the extra sudo invocation. I don't mind the iterative approach, myself.

**Comment by jiriks74** [comment]
I'd first solve this simple thing and think about the rest as that may take some time.

**Review by aclist — CHANGES_REQUESTED** [review]
```diff
0a1
>  local file="/proc/sys/vm/max_map_count"
2,3c3,4
<     if [[ ! -f /proc/sys/vm/max_map_count ]]; then
<         logger WARN "File \`/proc/sys/vm/max_map_count\` desn't exist!"
---
>     if [[ ! -f $file ]]; then
>         logger WARN "File '$file' doesn't exist!"
6c7
<     current_count=$(cat /proc/sys/vm/max_map_count)
---
>     current_count=$(cat "$file")
```

1. Read the filename into a variable so we just need to change one line here and not hardcode the path on subsequent lines
2. DZGUI uses the convention of single quotation marks inside of double quotation marks when printing a value. You don't have to escape backticks this way, and it's easier to read
3. Careful of typos: desn't -> doesn't

**Comment by jiriks74** [comment]
Hey there, sorry for taking few days, bad stuff that needed to be taken care of. I hope I'll be able to patch this in the afternoon.

**Review by aclist — APPROVED** [review]

**Comment by aclist** [comment]
> Hey there, sorry for taking few days, bad stuff that needed to be taken care of. I hope I'll be able to patch this in the afternoon.

Sorry to hear that. There is no rush.

I'll be unit and integration testing this and then approving it after I approve https://github.com/aclist/dztui/pull/177.

**Pull request merged**
